Client/Frontend/Browser/HomePanel/BackgroundImage.swift

```diff
@@ -0,0 +1,131 @@
+        if BackgroundImage.sponsors == nil && Preferences.NewTabPage.backgroundSponsoredImages.value {
+            BackgroundImage.sponsors = BackgroundImage.generateSponsoredData(file: sponsoredFilePath)
+        }
+        
+        self.info = BackgroundImage.randomBackground()
+    }
+    
+    private static func generateSponsoredData(file: String) -> [Background] {
+        guard let json = BackgroundImage.loadImageJSON(file: file),
+            let region = NSLocale.current.regionCode else {
+                return []
+        }
+        
+        let dateFormatter = DateFormatter().then {
+            $0.locale = Locale(identifier: "en_US_POSIX")
+            $0.dateFormat = "yyyy-MM-dd'T'HH:mm:ssZZZZZ"
+            $0.calendar = Calendar(identifier: .gregorian)
+        }
+        
+        let today = Date()
+        // Filter down to only regional supported items
+        let regionals = json.filter { ($0["regions"] as? [String])?.contains(region) == true }
+        
+        // Filter down to sponsors that fit the date requirements
+        let live = regionals.filter { item in
+            guard let dates = item["dates"] as? [String: String],
+                let start = dateFormatter.date(from: dates["start"] ?? ""),
+                let end = dateFormatter.date(from: dates["end"] ?? "") else {
+                    return false
+            }
+            
+            return today > start && today < end
+        }
+        
+        return generateBackgroundData(data: live)
+    }
+    
+    private static func generateStandardData(file: String) -> [Background] {
+        guard let json = BackgroundImage.loadImageJSON(file: file) else { return [] }
+        return generateBackgroundData(data: json)
+    }
+    
+    private static func generateBackgroundData(data: [[String: Any]]) -> [Background] {
+        return data.compactMap { item in
+            guard let image = item["image"] as? String,
+                let center = item["center"] as? CGFloat else { return nil }
+            
+            if let credit = item["credit"] as? [String: String],
+                let name = credit["name"] {
+                return Background(imageFileName: image, center: center, credit: (name, credit["url"]))
+            }
+            
+            return Background(imageFileName: image, center: center, credit: nil)
+        }
+    }
+    
+    private static func randomBackground() -> Background? {
+        // Determine what type of background to display
+        let sponsorshipShowRate = 4 // e.g. 4 == 25%
+        let useSponsor = hasSponsor && Int.random(in: 0..<sponsorshipShowRate) == 0
+        guard let dataSet = useSponsor ? sponsors : standardBackgrounds else { return nil }
+        if dataSet.count == 0 { return nil }
+        
+        let randomBackgroundIndex = Int.random(in: 0..<dataSet.count)
+        return dataSet[randomBackgroundIndex]
+    }
+    
+    private static func loadImageJSON(file: String) -> [[String: Any]]? {
```

Comment:
If the JSON file isn't a weird dynamic format, can we make it return a `Codable` structure?

Reply:
It isn't, however, it is extremely in-flux, meaning it will probably be complete redone during the next design iteration of this feature, so I just used dictionaries for now. Is this too 🤢🤮for you?